### DISCUSSION TITLE
readQuery always null

### DISCUSSION THREAD

**Discussion by XMadFoX** [urql-graphql/urql#2613] [Q&A] [answered]
Hello! I'm trying to do infinite scroll instead of pagination. So on new response I need to merge 2 arrays: existing items and new.
My query looks like:
```gql
{
  getItems(limit: 8, offset: 0) {
    items {
      ...
    }
    total
  }
}
```
There are other optional arguments, but for now I'm not using them in query.

As I've gotten it, I can get existing items by using cache.readQuery(...)
```ts
const cacheExchangeOptions: CacheExchangeOpts = {
  resolvers: {
    Query: {
      getItems(parent, args, cache, info) {
        const new = parent.getItems as any;
        
        // Temporary solution, otherwise I'm getting recursion error
        if (args.offset !== 8) return current; // this will work only for second "page" load.
        
        const existing = cache.readQuery({
          query: GetItemsDocument,
          variables: {
            limit: 8,
            offset: 0
            },
          });
          
        // and there should be smth like
        return {...current, items: [...existing.items, ,...current.items]}
      },
    },
  },
};
```
Same query as in `useQuery`, same variables (I used urql dev tools to verify), but `readQuery` result is always `null`.
Arguments from `inspectFields` for 1st page are also the same as variables in `readQuery`.

Any ideas what am I doing wrong?

**Comment by JoviDeCroock** [comment, marked as answer]
Could take inspiration from [our example resolver](https://github.com/FormidableLabs/urql/blob/main/exchanges/graphcache/src/extras/simplePagination.ts) and generally we also advise people that these are maybe better handled as UI concerns like [our repo example](https://github.com/FormidableLabs/urql/blob/main/examples/with-graphcache-pagination/src/PaginatedNpmSearch.jsx).

As to what you are doing wrong that might have something to do with the `return` or a non-nullable field being null, this resolver is getting triggered during readQuery so you might want to return keys rather than entities.

i.e.

```js
const firstPageKeys = cache.resolve('Query', 'getItems', firstPageVariables)
return firstPageKeys;
```

Then you are leaving the logic of resolving to actual entities to the read logic while just returning the keys for this root-field

**Reply by XMadFoX** [reply]
Thank you for feedback. I still haven't come up with how to do this with custom resolver (looks like it's not possible to return array of keys or get actual values by key to merge items inside), but yeah, I think I would do this logic at component/page level instead of modifing global behavoior.

**Reply by JoviDeCroock** [reply]
I mean, the first link is a working version
